### PR TITLE
bf16 support for per_channel bwd

### PR DESCRIPTION
**Pull request by liangel-02**
Follow up to #165098 - adding bf16 support for the backward pass. To avoid BC breaking changes/losing precision, we upcast the parameters to fp32 after the op gets called, and downcast the gradients to bf16 before returning.

For testing, we upcast to fp32 before calling the reference function. We increase the tolerance to 1e-2 for bf16 inputs because of a difference in casting calculations between python's `x.to(torch.bfloat16)` and cpp's `x.to(at::kBFloat16)` (after comparing intermediate tensors, we found that the numerics diverge after the final casting). We don't explicitly cast in the CPP op but rather let autograd/optimizer handle it.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #165362
* __->__ #165325

### CODE REVIEW

File: test/quantization/core/test_workflow_ops.py [hunk 1]
```diff
@@ -931,7 +943,9 @@ def _test_learnable_backward_per_channel(self, X_base, device, scale_base, zero_
                 dScale_actual = scale_curr.to(device).grad.detach()
                 dZeroPoint_expected = dZeroPoint.to(device).detach()
                 dZeroPoint_actual = zero_point_curr.to(device).grad.detach()
-                tolerance = 1e-4
+
+                # increasing tolerance for bf16 due to differences in python's x.to(torch.bfloat16) and cpp's x.to(at::kBFloat16)
```
Comment (andrewor14) [inline test/quantization/core/test_workflow_ops.py:948]: can you put an actual numerical example where the rounding differs? Like X is rounded to 0.168 in python but 0.167 in cpp or something?

File: test/quantization/core/test_workflow_ops.py [hunk 2]
```diff
@@ -152,6 +158,12 @@ def _fake_quantize_learnable_per_channel_affine_grad_reference(
 
         grad_scale[i] = grad_scale_i
         grad_zero_point[i] = grad_zp_i
+
+    if dtype is torch.bfloat16:
```
Comment (andrewor14) [inline test/quantization/core/test_workflow_ops.py:163]: I feel this needs a short comment, maybe something like mimic autograd downcasting to bf16, even though this part does not exist in the actual cpp op

**Review by andrewor14 — APPROVED** [review]
Looks great, thanks!
